cpp/src/fil/infer.cu

```diff
@@ -198,8 +201,46 @@ struct finalize_block {
     : tmp_storage(tmp_storage_), num_classes(num_classes_) {}
 
   template <int NITEMS>
-  static size_t smem_footprint() {
-    return sizeof(typename BlockReduceHostBestClass<NITEMS>::TempStorage);
+  static size_t smem_footprint(bool predict_proba) {
+    return predict_proba
+             ? sizeof(typename BlockReduceHost<NITEMS>::TempStorage)
+             : sizeof(typename BlockReduceHostBestClass<NITEMS>::TempStorage);
+  }
+
+  template <int NITEMS>
+  __device__ __forceinline__ vec<NITEMS, float> shifted_exp(
```

Comment:
It would be better to have a single-element version of this function, and use the `Vectorize` template above to apply it to `vec<NITEMS, float>`.